### ISSUE TITLE
[RFC] Upstream Chunked Prefill

### ISSUE THREAD

**Issue by rkooo567** [vllm-project/vllm#3130]
# Progress
- 2D -> 1D query refactoring https://github.com/vllm-project/vllm/issues/3130
- Sequence group metadata API update https://github.com/vllm-project/vllm/pull/3538/files
- Scheduler refactoring https://github.com/vllm-project/vllm/pull/3550
- Chunked prefill scheduler https://github.com/vllm-project/vllm/pull/3853
- Chunked prefill attention update  https://github.com/vllm-project/vllm/pull/3884
- logprob fix https://github.com/vllm-project/vllm/pull/4309
- Enable cuda-graph on prefill path -> Won't do

**Future Extension (not a scope of this RFC)**
- Make it work with sliding window attention
- Use same mechanism for prefix caching
- Use better kernel than context attention forward (ideally flash infer)

Chunked prefill chunks prefill requests into multiple chunks and batch them with decoding requests. Since prefill requests are compute-bound and decoding requests are memory-bound, it can greatly improve system efficiency by overlapping these two. See the linked papers for more details.
- https://arxiv.org/pdf/2401.08671.pdf
- https://arxiv.org/pdf/2308.16369.pdf 

We are planning to upstream chunked prefill internally used at Anyscale to OSS Vllm repo. We turned on this feature in the Anyscale production endpoint. 

# Benchmark Result
See https://github.com/vllm-project/vllm/issues/3130#issuecomment-2011281519

The following diagram is the benchmark result of Llama 13B x 2 A100 for different QPS (it is the result from Anyscale forked vLLM). Chunked prefill greatly improves latency when QPS is high, but has competitive performance at low QPS. The detailed benchmark results for different parameters with OSS vLLM is in progress (ETA end of this week). 

<img width="1496" alt="Screenshot 2024-03-01 at 11 30 35 AM" src="https://github.com/vllm-project/vllm/assets/18510752/39a5a500-f6df-4b29-a570-82d72aaeaf87">

# Design

## Kernel

For chunked prefill, we are internally using flash attn with paged attention enabled. ~~so we will also upstream flash attn integration first.~~ We decided to go with existing context attention first, and migrate to flash infer. ~~Note that the kernel choice is not finalized, and we are actively investigating other kernels like FlashInfer.~~

## Scheduler & Batch Layout

- Since chunked prefill mixes the prefill and decoding, having 2D query with batch dimensions is inefficient. To get around this, we will move 2D query back to 1D query. 
- We change Sequence so that it keeps track of the number prompting tokens have been prefilled.
- In Scheduler, if chunked prefill is enabled, it will first prioritize decoding requests, and then pick prompting request. If prompting request is too long, the prompting request will be chunked.
- The worker now will create metadata for both decoding requests and prompting requests, including their block tables, context lengths, and number of decoding and prefiling tokens and requests. 

```
    Chunked Prefill support:
    If chunked prefill is enabled, the input will include both prompt tokens
    and generation tokens. The layout is as follows:
    |<---------------------- num_valid_tokens -------------------------->|
    |<--------- num_prompt_tokens ----->|<--- num_generation_tokens >|
    |<-prompt_0->|<-prompt_1->|...............|<-gen_0->|<-gen_1->|..............|
    Notice that both num_valid_tokens, num_prompt_tokens and num_generation_tokens
    includes padding.
    The actual prompt length and offeset are stored in cum_prompt_context_lens.
    The actual num generation tokens are stored in num_generation_tokens_tensor.
    To support chunked prefill, where the prompt and context might have different
    length, we stored the context's length in cum_prompt_context_lens.
```
- Once request is processed and sampled, the result belongs to chunked prefilling requests will be ignored.

## Cuda Graph
Cuda graph won't be supported for chunked prefill because the performance benefit of cuda graph only happens with smaller # of tokens. 

~~Following is how we make everything cuda-graphable. We pad both prompt_tokens and generation tokens, and the cuda-graph is cached based number of prompt-tokens, number of generation tokens, and prompt batch size. Following are the details.~~

```
        # The key for caching cuda-graph.
        # We ensure that with the same total number of tokens,
        # same number of prompts, and same number of prompt tokens,
        # the intermediate tensors have the same shapes thus can be
        # replayed.
        #
        # Refer to following layout for the shapes of intermediate tensors.
        # during attention calculation.
        #
        #|<---------------------- padded_num_tokens ------------------------->|
        #|<--------- num_prompt_tokens ----->|<--- num_generation_tokens----->|
        #|<-prompt_0->|<-prompt_1->|...|<pad>|<-gen_0->|<-gen_1->|......|<pad>|
```

**Comment by tdene** [comment]
The "Benchmark Result" screenshot links directly to anyscale's private fork of vllm, and thus is not viewable by the general public.

**Comment by rkooo567** [comment]
@tdene thank you for pointing that out! I believe it is fixed

**Comment by rkooo567** [comment]
For kernel, we will try using `context_attention_forward` and then switch to flash infer once it is available

**Comment by rkooo567** [comment]
I got e2e working. Benchmarking result with OSS will be displayed very soon. I am also start focusing merging PRs starting from https://github.com/vllm-project/vllm/pull/3236.

**Comment by rkooo567** [comment]
Here's the initial benchmark result for chunked prefill. For the fairness of the benchmark, I used context attention in the default path.

I only had a throughput benchmark. As expected chunked prefill greatly improves the e2e throughput. Based on our internal benchmark, it works better as there is more load.

Note that chunked prefill has innately slower e2e latency as called out in this [blog](https://hao-ai-lab.github.io/blogs/distserve/#disaggregation-vs-chunked-prefill).
- ITL (inter token latency) as decoding tokens are batched with prefill tokens. We alleviated it by always batching decoding tokens first when scheduling.
- TTFT is supposed to be a bit slower because we chunk prefill (which adds per step overhead and KV cache memory load overhead).

We can alleviate this by not chunking prefill when there are no decoding requests.

In Anyscale’s internal benchmark we found it slower when QPS is 1 specifically. We will need more tuning and profiling for the optimal performance in this case.

The benchmark was performed on A100, 80GB, TP8, llama 70B
```
python benchmark_throughput.py --backend vllm --model "meta-llama/Llama-2-70b-hf" --dataset ../../data/ShareGPT_V3_unfiltered_cleaned_split.json --num-prompts 2000 -tp 8

Default OSS (on a 1dquery branch)
Throughput: 6.61 requests/s, 3198.02 tokens/s

Default OSS + use context attention (on a 1dquery branch)
Throughput: 6.51 requests/s, 3149.89 tokens/s

Chunk size 512 (cuda graph is decoding only)
Throughput: 7.72 requests/s, 3736.73 tokens/s

Chunk size 768 (cuda graph is decoding only)
Throughput: 7.51 requests/s, 3632.23 tokens/s
```

**Comment by njhill** [comment]
> In Anyscale’s internal benchmark we found it slower when QPS is 1 specifically. We will need more tuning and profiling for the optimal performance in this case.

@rkooo567 does this mean one request at a time i.e. batch size 1? I would expect there to be no latency impact for this case if we're not chunking prefill when there are no decoding requests? Or does QPS 1 still entail overlapping requests since they take > 1 sec each?

**Comment by rkooo567** [comment]
It means 1 request per 1 second. So if e2e takes more than 1 second, BS can be > 1. 

We haven't taken a deep look here (I think we will after this PR is merged to oss) for the root cause. But it could be becasue for longer prompt they are chunked when they don't have to be when QPS == 1.

**Comment by rkooo567** [comment]
Also, depending on scheudling strategy, you can optimize perf/throughput. For example, if you only allow to chunk 1 prompt per batch, it can actually improve ITL (because there's less interrupt from longer prompt for decoding). Or if you maximize chunk batches, it can maximize throughput (which is the benchmark above)

**Comment by rkooo567** [comment]
Update: these two tasks are going to be done as a follow up;

- Enable cuda-graph on prefill path
- Refactoring model runner

**Comment by rkooo567** [comment]
https://docs.google.com/document/d/1W6t6wouQKgl1QivS7gbkkY5xtR9Q6wvXbmOL-1onMmk/edit#heading=h.8um4c511b0b0

this is the initial benchmark result of chunked prefill!

TL;DR: chunked prefill improves ITL/e2e latency about 10~20% when QPS is sustainable. And almost 2X improvement at high QPS. It has small throughput improvement (less than 10%). It has comparable ITL at low QPS. However, it has higher TTFT compared to the default vLLM policy because the default vLLM policy is optimized for TTFT. There is still room for improvement such as enabling cuda graph for prefill or use better attention kernel for prefix prefill

**Comment by rkooo567** [comment]
<img width="1317" alt="Screenshot 2024-04-23 at 11 31 28 AM" src="https://github.com/vllm-project/vllm/assets/18510752/dc9df5e7-89ad-4da3-aa48-c89b24ce6bf0">
<img width="1314" alt="Screenshot 2024-04-23 at 11 31 25 AM" src="https://github.com/vllm-project/vllm/assets/18510752/0bb39e6a-73d5-470c-9ae8-01d40d6c8f69">

**Comment by mpjlu** [comment]
The performance of Chunked Prefill is better than no Chunk prefill with xformers. But when use flash-attn, with chunk prefill is worse than not use chunk prefill.
So do you have plan to use flash infer or other kernel for chunk prefill? @rkooo567 @WoosukKwon

**Comment by rkooo567** [comment]
Hmm @mpjlu can you tell me your benchmark environment? All of benchmark was done with flash-attn. 

It is possible it is not as fast as it could because the chunked prefill kernel is notably slower than flash attn. We are planning to migrate this, but the progress has been stuck for a while iiuc

**Comment by rkooo567** [comment]
Btw, the issue itself can be closed. All PRs are upstreamed and merged

**Comment by chqy99** [comment]
I saw that chunked-prefill improves ITL/e2e latency almost 2X improvement at high QPS from https://github.com/vllm-project/vllm/issues/3130. But in vllm-0.5.3 and two A100, I only almost improves 10~20%. Please tell me more test information, such as server-launch instruct 、client-request instruct、GPU setting el.

**Comment by chqy99** [comment]
Please summarize the changes in overall latency (TTFT + ITL). I tested it many times and the highest improvement was 18%.

**Comment by chqy99** [comment]
Please tell me how to set different QPS for experiments

**Comment by yiakwy-xpu-ml-framework-team** [comment]
in prefill stage, with chunk-prefill, vllm groups 16 prefill tokens together with priority in batches to compute with Flash attention (forward_prefix):
 
 ```
batch 0 ： decode  decode            decode
batch 1 ：         chunk-prefill     chunk-prefill
 ```
 
to reduce inter-token latency, imporve overall throughput with cost of slightly larger average of TTFT. Note, the forwrad_prefix is implemented with triton (triton style flash attention with KV cache).

![Screenshot 2024-08-15 191937](https://github.com/user-attachments/assets/7c489069-f388-42b9-a8d6-c8f0264c9c19)

While in default schedule all prefill tokens will be scheduled with priority so that you cannot batch 
 
 ```
batch 0 : docode decode                 decode
batch 1 :                P R E F I L L  decode
```

Where both TTFT and overal thoughput are worse than chunked-prefill.

So I guess @rkooo567 's test didn't relfect the truth. @chqy99, in summary, with chunked-prefill, your throughput will be improved around 17.5% (e.g., 4000 -> 4700)
